File: extensions/jetbrains-pymcu/src/main/kotlin/dev/begeistert/pymcu/boards/BoardRegistry.kt
// SPDX-License-Identifier: Apache-2.0
package dev.begeistert.pymcu.boards

/**
 * Catalog of all boards known to PyMCU, organized by manufacturer.
 * Flash sizes (flashKb) reflect available user flash after bootloader space,
 * matching the values displayed by PlatformIO.
 */
object BoardRegistry {

    val all: List<BoardEntry> = listOf(

        // ── Arduino ───────────────────────────────────────────────────────────

        BoardEntry("arduino_bt_mega168",        "Arduino BT ATmega168",                         "atmega168",   "Arduino", "avr",  16_000_000,  14, 1_024),
        BoardEntry("arduino_bt_mega328",        "Arduino BT ATmega328",                         "atmega328p",  "Arduino", "avr",  16_000_000,  28, 2_048),
        BoardEntry("arduino_duemil_mega168",    "Arduino Duemilanove / Diecimila ATmega168",    "atmega168",   "Arduino", "avr",  16_000_000,  14, 1_024),
        BoardEntry("arduino_duemil_mega328",    "Arduino Duemilanove / Diecimila ATmega328",    "atmega328p",  "Arduino", "avr",  16_000_000,  30, 2_048),
        BoardEntry("arduino_esplora",           "Arduino Esplora",                              "atmega32u4",  "Arduino", "avr",  16_000_000,  28, 2_048),
        BoardEntry("arduino_ethernet",          "Arduino Ethernet",                             "atmega328p",  "Arduino", "avr",  16_000_000,  31, 2_048),
        BoardEntry("arduino_fio",               "Arduino Fio",                                  "atmega328p",  "Arduino", "avr",   8_000_000,  30, 2_048),
        BoardEntry("arduino_industrial_101",    "Arduino Industrial 101",                       "atmega32u4",  "Arduino", "avr",  16_000_000,  28, 2_048),
        BoardEntry("arduino_leonardo",          "Arduino Leonardo",                             "atmega32u4",  "Arduino", "avr",  16_000_000,  28, 2_048),
        BoardEntry("arduino_lilypad_mega168",   "Arduino LilyPad ATmega168",                   "atmega168",   "Arduino", "avr",   8_000_000,  14, 1_024),
        BoardEntry("arduino_lilypad_mega328",   "Arduino LilyPad ATmega328",                   "atmega328p",  "Arduino", "avr",   8_000_000,  30, 2_048),
        BoardEntry("arduino_lilypad_usb",       "Arduino LilyPad USB",                         "atmega32u4",  "Arduino", "avr",   8_000_000,  28, 2_048),
        BoardEntry("arduino_mega",              "Arduino Mega",                                 "atmega2560",  "Arduino", "avr",  16_000_000, 248, 8_192),
        BoardEntry("arduino_mega_adk",          "Arduino Mega ADK",                             "atmega2560",  "Arduino", "avr",  16_000_000, 248, 8_192),
        BoardEntry("arduino_micro",             "Arduino Micro",                                "atmega32u4",  "Arduino", "avr",  16_000_000,  28, 2_048),
        BoardEntry("arduino_mini_mega168",      "Arduino Mini ATmega168",                       "atmega168",   "Arduino", "avr",  16_000_000,  14, 1_024),
        BoardEntry("arduino_mini_mega328",      "Arduino Mini ATmega328",                       "atmega328p",  "Arduino", "avr",  16_000_000,  28, 2_048),
        BoardEntry("arduino_nano_mega168",      "Arduino Nano ATmega168",                       "atmega168",   "Arduino", "avr",  16_000_000,  14, 1_024),
        BoardEntry("arduino_nano",              "Arduino Nano ATmega328",                       "atmega328p",  "Arduino", "avr",  16_000_000,  30, 2_048),
        BoardEntry("arduino_pro_mega168_3v3",   "Arduino Pro / Pro Mini ATmega168 (3.3V, 8MHz)","atmega168",  "Arduino", "avr",   8_000_000,  14, 1_024),
        BoardEntry("arduino_pro_mega168_5v",    "Arduino Pro / Pro Mini ATmega168 (5V, 16MHz)", "atmega168",  "Arduino", "avr",  16_000_000,  14, 1_024),
        BoardEntry("arduino_pro_mega328_3v3",   "Arduino Pro / Pro Mini ATmega328 (3.3V, 8MHz)","atmega328p", "Arduino", "avr",   8_000_000,  30, 2_048),
        BoardEntry("arduino_pro_mega328_5v",    "Arduino Pro / Pro Mini ATmega328 (5V, 16MHz)", "atmega328p", "Arduino", "avr",  16_000_000,  30, 2_048),
        BoardEntry("arduino_uno",               "Arduino Uno",                                  "atmega328p",  "Arduino", "avr",  16_000_000,  32, 2_048),
        BoardEntry("arduino_uno_r4_minima",     "Arduino Uno R4 Minima",                        "ra4m1",       "Arduino", "arm",  48_000_000, 256, 32_768),
        BoardEntry("arduino_uno_r4_wifi",       "Arduino Uno R4 WiFi",                          "ra4m1",       "Arduino", "arm",  48_000_000, 256, 32_768),
        BoardEntry("arduino_yun",               "Arduino Yun",                                  "atmega32u4",  "Arduino", "avr",  16_000_000,  28, 2_048),
        BoardEntry("arduino_zero",              "Arduino Zero",                                 "samd21g18a",  "Arduino", "arm",  48_000_000, 256, 32_768),

        // ── Atmel / Microchip — bare chips ────────────────────────────────────

        BoardEntry("atmega168_bare",   "ATmega168",   "atmega168",   "Atmel", "avr",  16_000_000,  16, 1_024),
        BoardEntry("atmega328p_bare",  "ATmega328P",  "atmega328p",  "Atmel", "avr",  16_000_000,  32, 2_048),
        BoardEntry("atmega32u4_bare",  "ATmega32U4",  "atmega32u4",  "Atmel", "avr",  16_000_000,  32, 2_560),
        BoardEntry("atmega88_bare",    "ATmega88",    "atmega88",    "Atmel", "avr",   8_000_000,   8, 1_024),
        BoardEntry("atmega2560_bare",  "ATmega2560",  "atmega2560",  "Atmel", "avr",  16_000_000, 256, 8_192),
        BoardEntry("attiny13_bare",    "ATtiny13",    "attiny13",    "Atmel", "avr",   9_600_000,   1,    64),
        BoardEntry("attiny44_bare",    "ATtiny44",    "attiny44",    "Atmel", "avr",   8_000_000,   4,   256),
        BoardEntry("attiny45_bare",    "ATtiny45",    "attiny45",    "Atmel", "avr",   8_000_000,   4,   256),
        BoardEntry("attiny84_bare",    "ATtiny84",    "attiny84",    "Atmel", "avr",   8_000_000,   8,   512),
        BoardEntry("attiny85_bare",    "ATtiny85",    "attiny85",    "Atmel", "avr",   8_000_000,   8,   512),
        BoardEntry("attiny2313_bare",  "ATtiny2313",  "attiny2313",  "Atmel", "avr",  20_000_000,   2,   128),
        BoardEntry("attiny4313_bare",  "ATtiny4313",  "attiny4313",  "Atmel", "avr",  20_000_000,   4,   256),

        // ── SparkFun ──────────────────────────────────────────────────────────

        BoardEntry("sparkfun_pro_micro_3v3",  "SparkFun Pro Micro 3.3V / 8MHz",  "atmega32u4", "SparkFun", "avr",  8_000_000,  28, 2_048),
        BoardEntry("sparkfun_pro_micro_5v",   "SparkFun Pro Micro 5V / 16MHz",   "atmega32u4", "SparkFun", "avr", 16_000_000,  28, 2_048),
        BoardEntry("sparkfun_mega_pro",       "SparkFun Mega Pro",               "atmega2560", "SparkFun", "avr", 16_000_000, 248, 8_192),
        BoardEntry("sparkfun_redboard",       "SparkFun RedBoard",               "atmega328p", "SparkFun", "avr", 16_000_000,  32, 2_048),

        // ── Adafruit ──────────────────────────────────────────────────────────

        BoardEntry("adafruit_trinket_3v3",  "Adafruit Trinket 3V / 8MHz",  "attiny85",  "Adafruit", "avr",  8_000_000,  8,  512),
        BoardEntry("adafruit_trinket_5v",   "Adafruit Trinket 5V / 16MHz", "attiny85",  "Adafruit", "avr", 16_000_000,  8,  512),
        BoardEntry("adafruit_gemma",        "Adafruit Gemma",              "attiny85",  "Adafruit", "avr",   8_000_000,  8,  512),
        BoardEntry("adafruit_flora",        "Adafruit Flora",              "atmega32u4","Adafruit", "avr",   8_000_000, 28, 2_048),
        BoardEntry("adafruit_metro_mega328","Adafruit Metro ATmega328",    "atmega328p","Adafruit", "avr",  16_000_000, 32, 2_048),
    )

    /** Boards grouped by manufacturer — drives the two-level tree. */
    val byManufacturer: Map<String, List<BoardEntry>> by lazy {
        all.groupBy { it.manufacturer }
            .toSortedMap()
    }

    fun findById(id: String): BoardEntry? = all.firstOrNull { it.id == id }

    fun findByChip(chip: String): List<BoardEntry> = all.filter { it.chip == chip }
}

File: extensions/vscode-pymcu/src/boards/boardRegistry.ts
export interface BoardEntry {
    id: string;
    name: string;
    chip: string;
    manufacturer: string;
    arch: string;
    frequency: number;
    flashKb: number;
    ramBytes: number;
}

export const BOARDS: BoardEntry[] = [
    // ── Arduino ──────────────────────────────────────────────────────────────
    { id: 'arduino_bt_mega168',       name: 'Arduino BT ATmega168',                          chip: 'atmega168',   manufacturer: 'Arduino',   arch: 'avr',    frequency: 16_000_000, flashKb: 14,  ramBytes: 1_024 },
    { id: 'arduino_bt_mega328',       name: 'Arduino BT ATmega328',                          chip: 'atmega328p',  manufacturer: 'Arduino',   arch: 'avr',    frequency: 16_000_000, flashKb: 28,  ramBytes: 2_048 },
    { id: 'arduino_duemil_mega168',   name: 'Arduino Duemilanove / Diecimila ATmega168',     chip: 'atmega168',   manufacturer: 'Arduino',   arch: 'avr',    frequency: 16_000_000, flashKb: 14,  ramBytes: 1_024 },
    { id: 'arduino_duemil_mega328',   name: 'Arduino Duemilanove / Diecimila ATmega328',     chip: 'atmega328p',  manufacturer: 'Arduino',   arch: 'avr',    frequency: 16_000_000, flashKb: 30,  ramBytes: 2_048 },
    { id: 'arduino_esplora',          name: 'Arduino Esplora',                               chip: 'atmega32u4',  manufacturer: 'Arduino',   arch: 'avr',    frequency: 16_000_000, flashKb: 28,  ramBytes: 2_048 },
    { id: 'arduino_ethernet',         name: 'Arduino Ethernet',                              chip: 'atmega328p',  manufacturer: 'Arduino',   arch: 'avr',    frequency: 16_000_000, flashKb: 31,  ramBytes: 2_048 },
    { id: 'arduino_fio',              name: 'Arduino Fio',                                   chip: 'atmega328p',  manufacturer: 'Arduino',   arch: 'avr',    frequency:  8_000_000, flashKb: 30,  ramBytes: 2_048 },
    { id: 'arduino_industrial_101',   name: 'Arduino Industrial 101',                        chip: 'atmega32u4',  manufacturer: 'Arduino',   arch: 'avr',    frequency: 16_000_000, flashKb: 28,  ramBytes: 2_048 },
    { id: 'arduino_leonardo',         name: 'Arduino Leonardo',                              chip: 'atmega32u4',  manufacturer: 'Arduino',   arch: 'avr',    frequency: 16_000_000, flashKb: 28,  ramBytes: 2_048 },
    { id: 'arduino_lilypad_mega168',  name: 'Arduino LilyPad ATmega168',                    chip: 'atmega168',   manufacturer: 'Arduino',   arch: 'avr',    frequency:  8_000_000, flashKb: 14,  ramBytes: 1_024 },
    { id: 'arduino_lilypad_mega328',  name: 'Arduino LilyPad ATmega328',                    chip: 'atmega328p',  manufacturer: 'Arduino',   arch: 'avr',    frequency:  8_000_000, flashKb: 30,  ramBytes: 2_048 },
    { id: 'arduino_lilypad_usb',      name: 'Arduino LilyPad USB',                          chip: 'atmega32u4',  manufacturer: 'Arduino',   arch: 'avr',    frequency:  8_000_000, flashKb: 28,  ramBytes: 2_048 },
    { id: 'arduino_mega',             name: 'Arduino Mega',                                  chip: 'atmega2560',  manufacturer: 'Arduino',   arch: 'avr',    frequency: 16_000_000, flashKb: 248, ramBytes: 8_192 },
    { id: 'arduino_mega_adk',         name: 'Arduino Mega ADK',                              chip: 'atmega2560',  manufacturer: 'Arduino',   arch: 'avr',    frequency: 16_000_000, flashKb: 248, ramBytes: 8_192 },
    { id: 'arduino_micro',            name: 'Arduino Micro',                                 chip: 'atmega32u4',  manufacturer: 'Arduino',   arch: 'avr',    frequency: 16_000_000, flashKb: 28,  ramBytes: 2_048 },
    { id: 'arduino_mini_mega168',     name: 'Arduino Mini ATmega168',                        chip: 'atmega168',   manufacturer: 'Arduino',   arch: 'avr',    frequency: 16_000_000, flashKb: 14,  ramBytes: 1_024 },
    { id: 'arduino_mini_mega328',     name: 'Arduino Mini ATmega328',                        chip: 'atmega328p',  manufacturer: 'Arduino',   arch: 'avr',    frequency: 16_000_000, flashKb: 28,  ramBytes: 2_048 },
    { id: 'arduino_nano_mega168',     name: 'Arduino Nano ATmega168',                        chip: 'atmega168',   manufacturer: 'Arduino',   arch: 'avr',    frequency: 16_000_000, flashKb: 14,  ramBytes: 1_024 },
    { id: 'arduino_nano',             name: 'Arduino Nano ATmega328',                        chip: 'atmega328p',  manufacturer: 'Arduino',   arch: 'avr',    frequency: 16_000_000, flashKb: 30,  ramBytes: 2_048 },
    { id: 'arduino_pro_mega168_3v3',  name: 'Arduino Pro / Pro Mini ATmega168 (3.3V, 8MHz)', chip: 'atmega168',   manufacturer: 'Arduino',   arch: 'avr',    frequency:  8_000_000, flashKb: 14,  ramBytes: 1_024 },
    { id: 'arduino_pro_mega168_5v',   name: 'Arduino Pro / Pro Mini ATmega168 (5V, 16MHz)',  chip: 'atmega168',   manufacturer: 'Arduino',   arch: 'avr',    frequency: 16_000_000, flashKb: 14,  ramBytes: 1_024 },
    { id: 'arduino_pro_mega328_3v3',  name: 'Arduino Pro / Pro Mini ATmega328 (3.3V, 8MHz)', chip: 'atmega328p',  manufacturer: 'Arduino',   arch: 'avr',    frequency:  8_000_000, flashKb: 30,  ramBytes: 2_048 },
    { id: 'arduino_pro_mega328_5v',   name: 'Arduino Pro / Pro Mini ATmega328 (5V, 16MHz)',  chip: 'atmega328p',  manufacturer: 'Arduino',   arch: 'avr',    frequency: 16_000_000, flashKb: 30,  ramBytes: 2_048 },
    { id: 'arduino_uno',              name: 'Arduino Uno',                                   chip: 'atmega328p',  manufacturer: 'Arduino',   arch: 'avr',    frequency: 16_000_000, flashKb: 32,  ramBytes: 2_048 },
    { id: 'arduino_yun',              name: 'Arduino Yun',                                   chip: 'atmega32u4',  manufacturer: 'Arduino',   arch: 'avr',    frequency: 16_000_000, flashKb: 28,  ramBytes: 2_048 },
    { id: 'arduino_zero',             name: 'Arduino Zero',                                  chip: 'samd21g18a',  manufacturer: 'Arduino',   arch: 'arm',    frequency: 48_000_000, flashKb: 256, ramBytes: 32_768 },
    // ── Atmel / Microchip — bare chips ────────────────────────────────────────
    { id: 'atmega168_bare',   name: 'ATmega168',   chip: 'atmega168',   manufacturer: 'Atmel', arch: 'avr', frequency: 16_000_000, flashKb: 16,  ramBytes: 1_024 },
    { id: 'atmega328p_bare',  name: 'ATmega328P',  chip: 'atmega328p',  manufacturer: 'Atmel', arch: 'avr', frequency: 16_000_000, flashKb: 32,  ramBytes: 2_048 },
    { id: 'atmega32u4_bare',  name: 'ATmega32U4',  chip: 'atmega32u4',  manufacturer: 'Atmel', arch: 'avr', frequency: 16_000_000, flashKb: 32,  ramBytes: 2_560 },
    { id: 'atmega88_bare',    name: 'ATmega88',    chip: 'atmega88',    manufacturer: 'Atmel', arch: 'avr', frequency:  8_000_000, flashKb:  8,  ramBytes: 1_024 },
    { id: 'atmega2560_bare',  name: 'ATmega2560',  chip: 'atmega2560',  manufacturer: 'Atmel', arch: 'avr', frequency: 16_000_000, flashKb: 256, ramBytes: 8_192 },
    { id: 'attiny13_bare',    name: 'ATtiny13',    chip: 'attiny13',    manufacturer: 'Atmel', arch: 'avr', frequency:  9_600_000, flashKb:  1,  ramBytes:    64 },
    { id: 'attiny44_bare',    name: 'ATtiny44',    chip: 'attiny44',    manufacturer: 'Atmel', arch: 'avr', frequency:  8_000_000, flashKb:  4,  ramBytes:   256 },
    { id: 'attiny45_bare',    name: 'ATtiny45',    chip: 'attiny45',    manufacturer: 'Atmel', arch: 'avr', frequency:  8_000_000, flashKb:  4,  ramBytes:   256 },
    { id: 'attiny84_bare',    name: 'ATtiny84',    chip: 'attiny84',    manufacturer: 'Atmel', arch: 'avr', frequency:  8_000_000, flashKb:  8,  ramBytes:   512 },
    { id: 'attiny85_bare',    name: 'ATtiny85',    chip: 'attiny85',    manufacturer: 'Atmel', arch: 'avr', frequency:  8_000_000, flashKb:  8,  ramBytes:   512 },
    { id: 'attiny2313_bare',  name: 'ATtiny2313',  chip: 'attiny2313',  manufacturer: 'Atmel', arch: 'avr', frequency: 20_000_000, flashKb:  2,  ramBytes:   128 },
    { id: 'attiny4313_bare',  name: 'ATtiny4313',  chip: 'attiny4313',  manufacturer: 'Atmel', arch: 'avr', frequency: 20_000_000, flashKb:  4,  ramBytes:   256 },
    // ── SparkFun ─────────────────────────────────────────────────────────────
    { id: 'sparkfun_pro_micro_3v3', name: 'SparkFun Pro Micro 3.3V / 8MHz',  chip: 'atmega32u4', manufacturer: 'SparkFun', arch: 'avr', frequency:  8_000_000, flashKb: 28,  ramBytes: 2_048 },
    { id: 'sparkfun_pro_micro_5v',  name: 'SparkFun Pro Micro 5V / 16MHz',   chip: 'atmega32u4', manufacturer: 'SparkFun', arch: 'avr', frequency: 16_000_000, flashKb: 28,  ramBytes: 2_048 },
    { id: 'sparkfun_mega_pro',      name: 'SparkFun Mega Pro',               chip: 'atmega2560', manufacturer: 'SparkFun', arch: 'avr', frequency: 16_000_000, flashKb: 248, ramBytes: 8_192 },
    { id: 'sparkfun_redboard',      name: 'SparkFun RedBoard',               chip: 'atmega328p', manufacturer: 'SparkFun', arch: 'avr', frequency: 16_000_000, flashKb: 32,  ramBytes: 2_048 },
    // ── Adafruit ─────────────────────────────────────────────────────────────
    { id: 'adafruit_trinket_3v3',   name: 'Adafruit Trinket 3V / 8MHz',   chip: 'attiny85',   manufacturer: 'Adafruit', arch: 'avr', frequency:  8_000_000, flashKb:  8, ramBytes:   512 },
    { id: 'adafruit_trinket_5v',    name: 'Adafruit Trinket 5V / 16MHz',  chip: 'attiny85',   manufacturer: 'Adafruit', arch: 'avr', frequency: 16_000_000, flashKb:  8, ramBytes:   512 },
    { id: 'adafruit_gemma',         name: 'Adafruit Gemma',               chip: 'attiny85',   manufacturer: 'Adafruit', arch: 'avr', frequency:  8_000_000, flashKb:  8, ramBytes:   512 },
    { id: 'adafruit_flora',         name: 'Adafruit Flora',               chip: 'atmega32u4', manufacturer: 'Adafruit', arch: 'avr', frequency:  8_000_000, flashKb: 28, ramBytes: 2_048 },
    { id: 'adafruit_metro_mega328', name: 'Adafruit Metro ATmega328',     chip: 'atmega328p', manufacturer: 'Adafruit', arch: 'avr', frequency: 16_000_000, flashKb: 32, ramBytes: 2_048 },
];

const _byManufacturer: Map<string, BoardEntry[]> = new Map();
for (const b of BOARDS) {
    let list = _byManufacturer.get(b.manufacturer);
    if (!list) { list = []; _byManufacturer.set(b.manufacturer, list); }
    list.push(b);
}

export const byManufacturer: ReadonlyMap<string, BoardEntry[]> =
    new Map([..._byManufacturer.entries()].sort(([a], [b]) => a.localeCompare(b)));

export function findById(id: string): BoardEntry | undefined {
    return BOARDS.find(b => b.id === id);
}

export function findByChip(chip: string): BoardEntry[] {
    return BOARDS.filter(b => b.chip === chip);
}

export function formatFrequency(hz: number): string {
    return hz >= 1_000_000 ? `${hz / 1_000_000} MHz` : `${hz / 1_000} kHz`;
}
